### PR TITLE
feat: message kinds

### PR DESCRIPTION
**Pull request by tydeu**
This PR adds a new definition `Message.kind` which returns the top-level tag of a message. This is serialized as the new field `kind` in `SerialMessaege` so that i can be used by external consumers (e.g., Lake) to identify messages via  `lean --json`.

The tag of trace messages has also been changed from `_traceMsg` to the more friendly `trace`.

### CODE REVIEW

**Comment by tydeu** [comment]
This is the first step for a potential solution to #5194.  An alternative solution would be to collect *all* tags within a message and serialize them. However, as the goal here is to be able to identify messages by name to customize reporting, it is not clear how this would work in a multiple tag case. Hence, the approach here is restricted to a single "message kind". A "tags" field can be added in the future should a more appropriate use case emerge.

**Comment by Kha** [comment]
If it's always inferred, why is it a field?

**Comment by tydeu** [comment]
@Kha It is a field so that it will be part of  `SerialMessage` (where it is not inferred as `data` is a `String`), and also thus a part of the JSON output. This solution seemed less disruptive than the alternative of different structures for `Message` and  `SerialMessage`.

**Comment by Kha** [comment]
Oh you mean it is there *for `deriving ToJson`*. I don't love that it can't be defaulted. What would be disruptive about putting it in `SerialMessage`, can't we make it a structure extending `BaseMessage String`?

**Comment by tydeu** [comment]
@Kha 

> What would be disruptive about putting it in `SerialMessage`, can't we make it a structure extending `BaseMessage String`?

Good idea! For some reason it did not occur to me that `SerialMessage` could be written that way.  (I was too stuck on them both needing to be `BaseMessage α` for some `α`.)

Separately, the odd JSON encoding of a anonymous kind has led me to propose #5980.

File: src/Lean/Message.lean [hunk 1]
```diff
@@ -307,8 +320,9 @@ structure BaseMessage (α : Type u) where
   endPos        : Option Position := none
   /-- If `true`, report range as given; see `msgToInteractiveDiagnostic`. -/
   keepFullRange : Bool := false
-  severity      : MessageSeverity := MessageSeverity.error
+  severity      : MessageSeverity := .error
   caption       : String          := ""
+
```
Comment (Kha) [inline src/Lean/Message.lean:325]: spurious newline

File: src/Lean/Message.lean [hunk 2]
```diff
@@ -320,7 +334,10 @@ abbrev Message := BaseMessage MessageData
 /-- A `SerialMessage` is a `Message` whose `MessageData` has been eagerly
 serialized and is thus appropriate for use in pure contexts where the effectful
 `MessageData.toString` cannot be used. -/
-abbrev SerialMessage := BaseMessage String
+structure SerialMessage extends BaseMessage String where
+  /-- The message kind (i.e., the top-level tag). -/
+  kind          : Name := .anonymous
```
Comment (Kha) [inline src/Lean/Message.lean:339]: I probably would have elided the default if there's no upside and the downside of accidentally forgetting to set it

**Review by Kha — COMMENTED** [review]
LGTM!

**Pull request merged**